src/libcore/ptr/const_ptr.rs

```diff
@@ -295,6 +295,62 @@ impl<T: ?Sized> *const T {
         intrinsics::ptr_offset_from(self, origin)
     }
 
+    /// Returns whether two pointers are guaranteed equal. Returns `true` whenever the
+    /// pointers are known to be equal. If the pointers are equal, but the platform
+    /// isn't sure the pointers are equal, this function returns `false`.
+    ///
+    /// This function is the mirror of [`guaranteed_ne`], but not its inverse. There are pointer
+    /// comparisons for which both functions return false.
+    ///
+    /// [`guaranteed_ne`]: #method.guaranteed_ne
+    ///
+    /// # Safety
+    ///
+    /// This function may not be used to change the behaviour of any code. It may
+    /// only be used for short circuiting operations out of performance reasons.
+    /// It is unsound for this function to be used to allow safe code to observe
```

Comment:
For this to be unsound, it has to violate some soundness condition. Our usual soundness condition is "you may not cause UB from safe code", which is unaffected by this. So I am not sure if soundness is the right term here. Your new soundness condition (which IMO requires t-lang FCP on its own) also seems to involve relational reasoning, which is concerning -- I'd prefer for soundness to be a unary property, i.e. one that can be evaluated by looking at individual execution traces (as opposed to a property that relates multiple execution traces).

Reply:
Right, I adjusted the comment and made the function safe